### PR TITLE
[LOOP-2358] [LOOP-995] Make pump sim error states affect insulin delivery

### PR DESCRIPTION
**Pull request by nhamming**
https://tidepool.atlassian.net/browse/LOOP-2358

mock pump error states now affect insulin delivery

### CODE REVIEW

File: MockKit/MockPumpManager.swift [hunk 1]
```diff
@@ -122,7 +125,7 @@ public final class MockPumpManager: TestingPumpManager {
                                                          imageName: "exclamationmark.circle.fill",
                                                          state: .critical)
         }
-        else if state.reservoirUnitsRemaining == 0 {
+        else if state.reservoirUnitsRemaining <= 0 {
```
Comment (darinkrauss) [inline MockKit/MockPumpManager.swift]: [nit] Should we consider just making sure the `reservoirUnitsRemaining` is pinned at zero?

File: MockKit/MockPumpManager.swift [hunk 2]
```diff
@@ -358,14 +374,22 @@ public final class MockPumpManager: TestingPumpManager {
     public func enactTempBasal(unitsPerHour: Double, for duration: TimeInterval, completion: @escaping (PumpManagerResult<DoseEntry>) -> Void) {
         logDeviceComms(.send, message: "Temp Basal \(unitsPerHour) U/hr Duration:\(duration.hours)")
 
-        if state.tempBasalEnactmentShouldError {
+        if state.tempBasalEnactmentShouldError || state.pumpBatteryChargeRemaining == 0 {
```
Comment (darinkrauss) [inline MockKit/MockPumpManager.swift:377]: Should there be a check for `reservoirUnitsRemaining`?

File: MockKit/MockPumpManager.swift [hunk 3]
```diff
@@ -402,14 +426,22 @@ public final class MockPumpManager: TestingPumpManager {
...
             completion(.failure(error))
         } else if state.deliveryCommandsShouldTriggerUncertainDelivery {
             state.deliveryIsUncertain = true
             logDeviceComms(.error, message: "Uncertain delivery for bolus")
             completion(.failure(PumpManagerError.uncertainDelivery))
+        } else if state.occlusionDetected || state.pumpErrorDetected {
+            let error = PumpManagerError.deviceState(MockPumpManagerError.pumpError)
+            logDeviceComms(.error, message: "Bolus failed because the pump is in an error state")
+            completion(.failure(error))
+        } else if state.reservoirUnitsRemaining <= 0 {
```
Comment (darinkrauss) [inline MockKit/MockPumpManager.swift]: Should there be a check for `case .suspended = state.suspendState`?

File: MockKit/MockPumpManager.swift [hunk 4]
```diff
@@ -213,7 +220,16 @@ public final class MockPumpManager: TestingPumpManager {
             if oldStatus != newStatus {
                 notifyStatusObservers(oldStatus: oldStatus)
             }
-
+            
+            // stop insulin delivery as pump state requires
+            if (newValue.occlusionDetected != oldValue.occlusionDetected && newValue.occlusionDetected) ||
+                (newValue.pumpErrorDetected != oldValue.pumpErrorDetected && newValue.pumpErrorDetected) ||
+                (newValue.pumpBatteryChargeRemaining != oldValue.pumpBatteryChargeRemaining && newValue.pumpBatteryChargeRemaining == 0) ||
+                (newValue.reservoirUnitsRemaining != oldValue.reservoirUnitsRemaining && newValue.reservoirUnitsRemaining <= 0)
```
Comment (darinkrauss) [inline MockKit/MockPumpManager.swift]: Should insulin delivery be stopped if `case .suspended = state.suspendState`?

**Review by darinkrauss — COMMENTED** [review]
A number of questions to consider.

**Comment by darinkrauss** [comment]
Just re-request a review when you've had a chance to consider these questions and respond. Thanks!

File: MockKit/MockPumpManager.swift (hunk 2 above)
Comment (nhamming) [reply MockKit/MockPumpManager.swift:377]: oh.... yes. will add

File: MockKit/MockPumpManager.swift (hunk 3 above)
Comment (nhamming) [reply MockKit/MockPumpManager.swift]: there is a bit further down.

File: MockKit/MockPumpManager.swift (hunk 4 above)
Comment (nhamming) [reply MockKit/MockPumpManager.swift]: It already gets stopped when the state is set to suspended.

File: MockKit/MockPumpManager.swift (hunk 1 above)
Comment (nhamming) [reply MockKit/MockPumpManager.swift]: yes, I was thinking this as well. I'll have a look at what happens.

**Review by darinkrauss — APPROVED** [review]
LGTM! 🚢

**Pull request merged**